Re-implement bundle dialogs as editor extensions

vlaaad opened the pull request:
The bundle dialogs now use newer UI, and are implemented as editor extensions.

In addition to that, we added some new features to editor scripts:
- commands can now use a new location: `"Bundle"` (menu items are added to **Project -> Bundle menu**)
- commands can now define another query parameter in addition to `selection`: `argument`. Currently, command arguments are only passed to the commands in the `"Bundle"` location. There, the argument is `true` if the user asked to configure the bundle options, and `false` otherwise (i.e. when re-bundling).
- commands can now define an `id`: an identifier that may be used to refer to the command by the editor. Currently, it's only used to remember the last used bundle command for re-bundling.
- we added `editor.open_external_file` function
- we added `json` module, with the usual `encode` and `decode` functions

Technical notes:

There are some minor changes to how the new bundling process invokes bob:
- we no longer include `--auth` and `--email` args that were always set by the editor to empty strings. The args are no longer required
- we no longer provide `--defoldsdk` arg, as it's set by bob to the same value that the editor would have provided

Fixes #8062

vlaaad committed:
Re-implement bundle dialogs as editor extensions

The bundle dialogs now use newer UI, and are implemented as editor extensions.

In addition to that, we added some new features to editor scripts:
- commands can now use a new location: `"Bundle"` (menu items are added to **Project -> Bundle menu**)
- commands can now define another query parameter in addition to `selection`: `argument`. Currently, command arguments are only passed to the commands in the `"Bundle"` location. There, the argument is `true` if the user asked to configure the bundle options, and `false` otherwise (i.e. when re-bundling).
- commands can now define an `id`: an identifier that may be used to refer to the command by the editor. Currently, it's only used to remember the last used bundle command for re-bundling.
- we added `editor.open_external_file` function
- we added `json` module, with the usual `encode` and `decode` functions

Technical notes:

There are some minor changes to how the new bundling process invokes bob:
- we no longer include `--auth` and `--email` args that were always set by the editor to empty strings. The args are no longer required
- we no longer provide `--defoldsdk` arg, as it's set by bob to the same value that the editor would have provided

Fixes #8062

vlaaad commented on editor/src/clj/editor/adb.clj:
```diff
@@ -1,118 +0,0 @@
-;; CONDITIONS OF ANY KIND, either express or implied. See the License for the
-;; specific language governing permissions and limitations under the License.
-
-(ns editor.adb
```
Deleted, now the functionality is implemented directly in the bundle editor script.

vlaaad commented on editor/src/clj/editor/app_view.clj:
```diff
@@ -2536,116 +2527,33 @@ If you do not specifically require different script states, consider changing th
+    (let [{:keys [command handler+context]} user-data]
+      (prefs/set! prefs [:bundle :last-bundle-command] (if (handler/synthetic-command? command) nil command))
+      (when (handler/enabled? handler+context)
+        (handler/run handler+context)))))
```
The `:bundle` command is now a "meta-command" — its work is presenting and running other commands that live in `:editor.bundle/menu` location.

vlaaad commented on editor/src/clj/editor/bundle.clj:
```diff
@@ -1,38 +0,0 @@
-;; CONDITIONS OF ANY KIND, either express or implied. See the License for the
-;; specific language governing permissions and limitations under the License.
-
-(ns editor.bundle
```
Deleted: the functionality here moved to the ios bundling section of `bundle.editor_script`.

vlaaad commented on editor/src/clj/editor/bundle_dialog.clj:
```diff
@@ -1,1003 +0,0 @@
-;; CONDITIONS OF ANY KIND, either express or implied. See the License for the
-;; specific language governing permissions and limitations under the License.
-
-(ns editor.bundle-dialog
```
Deleted: all functionality is re-implemented in `bundle.editor_script`

vlaaad commented on editor/src/clj/editor/code/view.clj:
```diff
@@ -2970,7 +2970,7 @@
    {:command :edit-breakpoint :label "Edit Breakpoint"}])
 
-(handler/register-menu! ::menubar :editor.app-view/view-end
+(handler/register-menu! ::menubar-view :editor.app-view/view-end
```
I had to refactor `editor.handler` ns to make built-in and extension commands more similar (e.g. to support command ids). Because of that, id handling was changed, and now you need separate registration ids (first arg to `handler/register-menu!`) for separate things. Previously, the id was scoped to location (`:editor.app-view/view-end` or `:editor.app-view/edit-end`). Now it is scoped to the handler system. Because of that, this ns has to use different registration ids for different locations.

vlaaad commented on editor/src/clj/editor/editor_extensions.clj:
```diff
@@ -368,6 +372,58 @@
+      ([{:keys [rt]} lua-string lua-options]
+       (decode lua-string (rt/->clj rt json-decode-options-coercer lua-options))))))
+
+(def json-value-coercer
+  (coerce/recursive
+    #(coerce/one-of
+       coerce/null
+       coerce/boolean
+       coerce/number
+       coerce/string
+       (coerce/vector-of % :min-count 1)
+       (coerce/map-of coerce/string %))))
+
+(def ext-json-encode
+  (rt/lua-fn ext-json-decode
+    ([{:keys [rt]} lua-value]
+     (json/write-str (rt/->clj rt json-value-coercer lua-value)))))
```
For `ios-deploy`, we need to `decode` json. I added `encode` for completeness, even though the editor scripts do not use it.

vlaaad commented on editor/src/clj/editor/editor_extensions.clj:
```diff
@@ -601,7 +665,8 @@
                :out (line-writer #(display-output! :out %))
                :err (line-writer #(display-output! :err %))
-               :env {"editor" {"get" (make-ext-get-fn project)
+               :env {"editor" {"bundle" {"project_binary_name" ext-project-binary-name} ;; undocumented, hidden API!
```
`editor.bundle` module is internal. It exists in the runtime, but the docs/completions are displayed only when running the editor in dev. This will allow us to change the bundling in the future if we decide to change the bundling UI/UX.

vlaaad commented on editor/src/clj/editor/editor_extensions/graph.clj:
```diff
@@ -121,14 +122,17 @@
-        #(some-> (properties/value outline-property) to)))))
+  (or (when (fn/multi-responds? ext-get node-id property evaluation-context)
+        #(ext-get node-id property evaluation-context))
+      (when (and (= :editor.game-project/GameProjectNode (node-id->type-keyword node-id evaluation-context))
+                 (re-matches #"^[a-zA-Z_][a-zA-Z0-9_]*\.[a-zA-Z_][a-zA-Z0-9_]*$" property))
+        #(game-project/get-setting node-id (string/split property #"\." 2) evaluation-context))
```
Now we can do e.g. `editor.get("/game.project", "project.title")` — we do this in `bundle.editor_script`

vlaaad commented on editor/src/clj/editor/editor_extensions/ui_components.clj:
```diff
@@ -494,6 +494,7 @@
     {:fx/type ext-with-extra-check-box-props
      :desc (-> {:fx/type fx.check-box/lifecycle
                 :style-class ["check-box" "ext-check-box"]
+                :mnemonic-parsing false
```
Somewhere it bit me... maybe in a check-box that had a label like `x86_64-linux`

vlaaad commented on editor/src/clj/editor/handler.clj:
```diff
@@ -19,197 +19,264 @@
+(defn synthetic-command? [x]
+  (= (namespace x) synthetic-command-str))
+
+(defn- register [state registration menus handlers]
+  (let [handlers (mapv (fn [h]
+                         (if (contains? h :command)
+                           h
+                           (assoc h :command (keyword synthetic-command-str (str synthetic-command-str (RT/nextID))))))
+                       handlers)
+        menus (reduce
+                (fn [acc [location menu-item]]
+                  (update acc location (fnil conj []) menu-item))
+                menus
+                (e/mapcat
+                  (fn [{:keys [command locations]}]
+                    (let [menu-item {:command command}]
+                      (e/map #(coll/pair % menu-item) locations)))
+                  handlers))]
+    (-> state
+        (unregister registration)
+        (update :menus add-menus menus registration)
+        (update :handlers add-handlers handlers registration)
+        (update :registrations assoc registration {:locations (vec (keys menus))
+                                                   :command+contexts (vec (e/mapcat
+                                                                            (fn [{:keys [command contexts]}]
+                                                                              (e/map #(coll/pair command %) contexts))
+                                                                            handlers))}))))
```
I refactored handler registration so it has more approachable semantics w.r.t. registration identifier, that were previously scoped differently in different places. Now there is a single `registration` id that is used to register both menus and handlers.

vlaaad commented on editor/src/clj/editor/handler.clj:
```diff
@@ -274,14 +341,6 @@
               handler))
           (vals (get-in state [:handlers [command ctx-name]])))))
 
-(defn- get-active [command command-contexts user-data evaluation-context]
-  (let [state @state-atom]
-    (some (fn [ctx]
-            (let [full-ctx (assoc-in ctx [:env :user-data] user-data)]
-              (when-let [handler (get-active-handler state command full-ctx evaluation-context)]
-                [handler full-ctx])))
-          command-contexts)))
```
inlined in `active`

vlaaad commented on editor/src/clj/editor/ios_deploy.clj:
```diff
@@ -1,123 +0,0 @@
-;; CONDITIONS OF ANY KIND, either express or implied. See the License for the
-;; specific language governing permissions and limitations under the License.
-
-(ns editor.ios-deploy
```
Deleted: the functionality is now implemented in the ios section of `bundle.editor_script`

vlaaad commented on editor/resources/prelude.lua:
```diff
@@ -99,4 +101,282 @@ editor.ui.resource_field = editor.ui.component(function(props)
+    return config
+end
+
+function editor.bundle.output_directory(requested_dialog, output_subdir)
```
The "get output directory" step. Translated from now-deleted `editor.bundle-dialog` ns.

vlaaad commented on editor/resources/prelude.lua:
```diff
@@ -99,4 +101,282 @@ editor.ui.resource_field = editor.ui.component(function(props)
+    return output_directory
+end
+
+function editor.bundle.create(config, output_directory, extra_bob_opts)
```
The "invoke bob" step. Mostly corresponds to generic bundle options that are now removed from `editor.pipeline.bob` ns.

vlaaad committed:
Fix build report creation regression

vlaaad committed:
Fix output directory creation on windows

matgis commented on editor/src/clj/editor/app_view.clj:
```diff
@@ -2536,116 +2527,33 @@ If you do not specifically require different script states, consider changing th
+                      :label (handler/label handler+context)
+                      :user-data {:command command
+                                  :handler+context handler+context}}))))
+             (into [])))))
```
Use transducer with `into` instead of creating an eduction?

Or you could make use of the new `coll/transfer` function if you prefer the added clarity.

matgis commented on editor/src/clj/editor/editor_extensions.clj:
```diff
@@ -368,6 +372,58 @@
+
+(def ext-json-decode
+  (letfn [(decode [^LuaString lua-string opts]
+            (with-open [r (-> lua-string .toInputStream io/reader (PushbackReader. 64))]
```
Is 64 bytes enough of a buffer?

matgis commented on editor/src/clj/editor/editor_extensions/commands.clj:
```diff
@@ -82,53 +82,58 @@
-                               (rt/->clj rt coerce/to-boolean (rt/invoke-immediate-1 (:rt state) active (rt/->lua opts) (:evaluation-context env)))))))
+    (cond-> {:contexts contexts
+             :label label
+             :locations locations}
```
Style: Missing newline?

matgis commented on editor/test/editor/prefs_test.clj:
```diff
@@ -231,6 +231,14 @@
+      (is (false? (prefs/set? p [])))
+      (prefs/set! p [] "new-string")
+      (is (true? (prefs/set? p []))))))
+
```
Maybe add tests for entries below some path as well?

vlaaad replied:
Good idea, I'll add!

matgis reviewed (changes requested):
Looks good! I think I may have found a place where a preference is read from the old unmigrated location.

I also have some suggestions for moving some Lua helper functions into utility modules, but that can wait.

vlaaad commented on editor/resources/bundle.editor_script:
```diff
@@ -0,0 +1,587 @@
+    })
+end)
+
+local function bundle_linux(requested_dialog)
```
Good idea, I'll rename the parameter

vlaaad commented on editor/resources/bundle.editor_script:
```diff
@@ -0,0 +1,587 @@
+    return nil
+end
+
+---@return string cmd ios-deploy command path
```
Yeah, I was using them for myself when writing the code, by they are not really needed. I'll remove them.

vlaaad commented on editor/resources/bundle.editor_script:
```diff
@@ -0,0 +1,587 @@
+        if editor.external_file_attributes(prefs_path).exists then
+            return prefs_path
+        else
+            print(("ios-deploy path defined in preferences does not exist: '%s'"):format(prefs_path))
```
These end up in the console. Maybe we want to introduce a `log` function that will print to editor logs (in addition to console? or instead of?), but for now I think writing to console with print is good enough.

vlaaad commented on editor/resources/bundle.editor_script:
```diff
@@ -0,0 +1,587 @@
+        if config.sign then
+            if #config["provisioning-profile"] == 0 then
+                ret.mobileprovisioning = "Provisioning profile must be set to sign the application."
+            elseif config["provisioning-profile"]:sub(-#".mobileprovision") ~= ".mobileprovision" then
```
Maybe, though I think that would require a design phase to also add some more goodies to it (like basename, all extensions etc.)

vlaaad commented on editor/resources/bundle.editor_script:
```diff
@@ -0,0 +1,587 @@
+
+--- ANDROID
+
+
```
Good catch!

vlaaad commented on editor/resources/bundle.editor_script:
```diff
@@ -0,0 +1,587 @@
+function M.get_commands()
+    return {
+        editor.bundle.command("iOS Application...", "bundle-ios", bundle_ios, {
+            active = function() return editor.platform:sub(-#"macos") == "macos" end
```
Great idea, I'll do it!

vlaaad commented:
@matgis I didn't quite understand your concerns about reading prefs from the old unmigrated location. If this is about `build.texture-compression`, we migrated it from the old prefs and it's still in `prefs.clj` built-in schema definition. I removed `bundle` (not `build`) prefs definitions from `prefs.clj` and moved them to `bundle.editor_script` since this is where they are used:
https://github.com/defold/defold/blob/71c54eeebe228f868de55d16a791278bffacc9f0/editor/resources/bundle.editor_script#L533-L540

vlaaad committed:
Review fixes

vlaaad committed:
Cmd.exe does not like forward slashes

matgis reviewed (approved):
Ship it!
